File: Έγγραφα SEM4105/Λύσεις ασκήσεων/09. Ένατο εργαστήριο/1/trapez.cpp
#include <iostream>
#include <cmath>
#include <cstddef>
#include <functional>


double g(double x)
{
    return std::sin(x);
}


double trapez(std::function<double(double)> f,
	      double a, double b, std::size_t n)
{
    double const h{(b-a)/n};

    double s{(f(a)+f(b))/2.0};

    for (std::size_t i{1}; i < n; ++i) {
	s += f(a+i*h);
    }

    return s*h;
}


int main()
{
    double constexpr pi{3.14159265358979323846};

    std::size_t n{2};

    double constexpr correct = 2.0;
    for (std::size_t k{1}; k < 10; ++k) {
        double integr = trapez(g, 0.0, pi, n);

	std::cout << n << ' ' << integr << ' '
		  << correct - integr << '\n';
	n*=2;
    }
    
}

File: Έγγραφα SEM4105/Λύσεις ασκήσεων/09. Ένατο εργαστήριο/1/trapez.py
import math



def g(x):
    return math.sin(x)


def trapez(f, a,b,n):
    
    h = (b-a)/n
    s = (f(a)+f(b))/2.0

    for i in range(1,n):
        s += f(a+i*h)
    
    return s*h


def main():
    n = 2
    correct = 2.0
    for k in range(1,10):
        integr = trapez(g, 0.0, math.pi, n)
        print(n, integr, correct - integr)
        n *= 2




main()
